docs/firmware/golioth-firmware-sdk/light-db-state/guide-lightdb-get.md

```diff
@@ -0,0 +1,142 @@
+enqueued by the Golioth Client so that program execution is not blocked. A
+user-defined callback function is supplied so that you may process the received
+data once the get operation is completed. An optional callback argument can be
+specified, this data will be available in the callback function.
+
+In the above example, we request the `"counter"` path from Golioth, specifying
+`counter_set_handler` as the callback function and no callback argument
+(`NULL`).
+
+### Callback function
+
+Callback functions need to follow the type specified by the function that
+registers them. In this case a `golioth_get_cb_fn` type callback function must
+be defined:
+
+```c
+static void counter_get_handler(golioth_client_t client,
+                                const golioth_response_t* response,
+                                const char* path,
+                                const uint8_t* payload,
+                                size_t payload_size,
+                                void* arg) {
+    if (response->status != GOLIOTH_OK) {
+        // The get operation failed.
+        return;
+    }
+
+    int32_t received_value = golioth_payload_as_int(payload, payload_size);
+
+    // Do something with the received_value variable
+    return;
+}
+```
+
+The response is passed to the callback function as a `golioth_response_t` struct
+that includes `status`. It is recommended that callbacks test for the
+`GOLIOTH_OK` status which indicates a successful operation.
+
+By default the data received from Golioth is in JSON format, stored in memory at
+the `payload` pointer location, with `payload_size` indicating the length of
+data. Helper functions like `golioth_payload_as_int()` are used to convert the
+payload to the desired type, or a JSON parser of your choosing may be used to
+unpack the payload.
+
+The same callback may be used by multiple async functions. Add logic to react
+based on which `path` was being written to, and any callback arguments found in
+the `arg` variable.
+
+## Getting Complex JSON Objects
+
+Both examples above get `int` type data from Golioth, and similar functions exist
+for `bool`, `float`, and `string`. However, it is often more convenient to get
+several paths in one operation. To do so, the API supports JSON objects.
+
+```c
+uint8_t json_buf[128];
+
+/* Get root of LightDB State, but JSON can be returned for any path */
+int err = golioth_lightdb_get_json_sync(client,
+                                        "",
+                                        json_buf,
+                                        sizeof(json_buf),
+                                        1);
+```
+
+In the example above, the `json_buf[128]` character array of an arbitrary length
+is created to store the returned a JSON object. The API call instructs Golioth
+to get the `""` (root) path data and store it in `json_buf` which has a length
+limit of `strlen(json_buf)`. This is an async function, so a timeout of `1`
```

Comment:
```suggestion
limit of `sizeof(json_buf)`. This is an async function, so a timeout of `1`
```
This brings up another question though - do you need to account for the trailing `\0`? I.e. should you pass `sizeof(json_buf) - 1` as the length?

Reply:
@sam-golioth I think this should be `strlen()` in both places, so line 117 should have been:

```c
int err = golioth_lightdb_get_json_sync(client,
                                        "",
                                        json_buf,
                                        strlen(json_buf),
                                        1);
```
`sizeof` is the actual size of the array at compile time. What we need is the number of characters before the null terminator and we'll only know that after `snprintf`/`snprintk` executes during runtime.

> do you need to account for the trailing \0? I.e. should you pass sizeof(json_buf) - 1 as the length?
```
json_str_len | Length of json_str, not including NULL terminator
```
According to our Doxygen, no. But I suspect you know that and are asking about adding a mention in the docs as part of this PR. :+1:

Reply:
This is a `get()` function, so there is nothing in `json_buf` and so `strlen()` won't work. The Doxygen comment you quoted is also form the `set()` function. It's not clear to me whether or not you need to account for the trailing `\0` in the `get()` functions.

Reply:
Ah thanks for catching this. I was confused because this discrepancy existed on both the `set` and `get` pages in this PR:

* `set` changed to use `strlen` and mention the null terminator
* `get` changed to use `sizeof`

> It's not clear to me whether or not you need to account for the trailing \0 in the get() functions.

Ah, I see. I don't know either. Looking at the code didn't clear it up. I think I need to test it both ways and see what happens.